component/azstorage/block_blob.go

```diff
@@ -1068,119 +1044,246 @@ func (bb *BlockBlob) TruncateFile(name string, size int64) error {
+				log.Err("BlockBlob::expandFile : Failed to stage block for %s [%s]", name, err.Error())
 				return err
 			}
 		}
+
+		blkList = append(blkList, id)
+		size -= blockSize
+	}
+
+	err := bb.CommitBlocks(blobName, blkList)
+	if err != nil {
+		log.Err("BlockBlob::expandFile : Failed to commit blocks for %s [%s]", name, err.Error())
+		return err
 	}
 
 	return nil
 }
+func (bb *BlockBlob) updateBlocks(name string, newSize int64, blockSize int64, blockList *internal.CommittedBlockList) error {
+	// Iterate through blockList to figure out how many blocks we can keep
+	removeBlocks := false
+	blkList := make([]string, 0)
 
-func (bb *BlockBlob) HandleSmallFile(name string, size int64, originalSize int64) ([]byte, error) {
-	var data = make([]byte, size)
-	var err error
-	if size > originalSize {
-		err = bb.ReadInBuffer(name, 0, 0, data)
+	blobName := filepath.Join(bb.Config.prefixPath, name)
+	blobClient := bb.Container.NewBlockBlobClient(blobName)
+
+	var block internal.CommittedBlock
+	lastIdx := 0
+
+	// Extract the block-id length and size of block so that we create new blocks accordingly
+	idLen := int64(16)
+	orgId, err := base64.StdEncoding.DecodeString((*blockList)[0].Id)
+	if err == nil {
+		idLen = int64(len(orgId))
+	}
+	oldBlockSize := int64((*blockList)[0].Size)
+
+	if oldBlockSize != 0 {
+		blockSize = oldBlockSize
+	}
+
+	for lastIdx, block = range *blockList {
+		if newSize < int64(block.Size) {
+			removeBlocks = true
+			break
+		}
+		newSize -= int64(block.Size)
+		blkList = append(blkList, block.Id)
+	}
+
+	if newSize == 0 {
+		// New Size exactly aligns to the block boundary
+		// Commit the current list and no need to upload any data here
+		err := bb.CommitBlocks(blobName, blkList)
 		if err != nil {
-			log.Err("BlockBlob::TruncateFile : Failed to read small file %s", name, err.Error())
+			log.Err("BlockBlob::updateBlocks : Failed to commit blocks for %s [%s]", name, err.Error())
 		}
-	} else {
-		err = bb.ReadInBuffer(name, 0, size, data)
+		return err
+	}
+
+	if removeBlocks {
+		// Last block in file needs truncation
+		data := make([]byte, newSize)
+
+		id := base64.StdEncoding.EncodeToString(common.NewUUIDWithLength(idLen))
+		err := bb.ReadInBuffer(name, (*blockList)[lastIdx].Offset, newSize, data)
+		if err != nil {
+			log.Err("BlockBlob::updateBlocks : Failed to read buffer for %s [%s]", name, err.Error())
+			return err
+		}
+
+		_, err = blobClient.StageBlock(context.Background(),
+			id,
+			streaming.NopCloser(bytes.NewReader(data)),
+			&blockblob.StageBlockOptions{
+				CPKInfo: bb.blobCPKOpt,
+			})
+
 		if err != nil {
-			log.Err("BlockBlob::TruncateFile : Failed to read small file %s", name, err.Error())
+			log.Err("BlockBlob::updateBlocks : Failed to stage block for %s [%s]", name, err.Error())
+			return err
 		}
+
+		blkList = append(blkList, id)
+		newSize = 0
 	}
-	return data, err
+
+	return bb.createSparseFile(name, newSize, blockSize, blkList, idLen)
```

Comment:
If file size is not aligning with the block size and truncate is used to expand the size, this will lead to irregular blocksizes in the block list.
Eg: lets say blocksize = 4M
file size = 5M (i.e., 2blocks) block1 = 4M, block2 = 1M

Call for truncate(10M)

I suppose the logic in this PR will make the block list look like this after truncate operation: 
block1 = 4M, block2 = 1M, block3 = 4M, block4 = 1M. 
Although this looks correct but the subsequent operations in block cache cannot utilize this sort of blocklist.
Please correct me if I am wrong